### ISSUE TITLE
Font Texture allocates too much memory for Unicode-supported (Chinese) fonts.

### ISSUE THREAD

**Issue by JX-Master** [ocornut/imgui#1859]
**Version/Branch of Dear ImGui:**

1.60

**Back-end file/Renderer/OS:** 

Windows 10 Direct3D 11.0

**My Issue/Question:** 

When we load a Unicode font TTF file using `AddFontFromFileTTF(GetGlyphRangesChinese)`, the program creates a texture with a full set of Chinese characters, which **consumes 200+MB memory only for displaying fonts**, while a non-Unicode fonts will only consume about 10MB memory.

This is not a bug though, but since we will only use **1% part of Chinese characters in character set**, I wonder if there is some way to optimize the memory allocation so that we don't create the texture for the whole character set, but only for the character that we really need. There are about 60k characters in Chinese Unicode fonts and much more for a multi-language font, this will be a problem when we have a memory budget.

**Screenshots/Video** 
Below is the font texture that from the font Deng.ttf(等线), which is a 4k x 8k texture that ImGui created.

<img width="1916" alt="capture" src="https://user-images.githubusercontent.com/25875303/40951677-9dc44a9c-68aa-11e8-9ffc-1d8a36025455.PNG">

**Comment by dpethes** [comment]
You can create your own character range instead of using GetGlyphRangesChinese, wouldn't that work for you?

**Comment by ocornut** [comment]
There is a GlyphRangeBuilder for the purpose of creating custom glyph ranges from source strings you can use. At the moment the font atlas needs to be statically created so that is the easiest solution.

**Comment by JX-Master** [comment]
Thanks, I have made a table of 2,500 most frequently used Simplified Chinese characters and encapsulated everything into a single function `const ImWchar*  ImFontAtlas::GetGlyphRangesSimplifiedChinese()` to solve this problem, which is pasted below:
```
const ImWchar*  ImFontAtlas::GetGlyphRangesSimplifiedChinese()
{
	// Store 2,500 regularly used Simplified Chinese characters in "Table of regularly used characters in contemporary Chinese Language".
	// Reference: https://zh.wiktionary.org/wiki/%E9%99%84%E5%BD%95:%E7%8E%B0%E4%BB%A3%E6%B1%89%E8%AF%AD%E5%B8%B8%E7%94%A8%E5%AD%97%E8%A1%A8
	// This table covers 97.97% of all characters used during the mouth in July, 1987.
	// If there are any characters missing, just look up the Unicode manually and append below.
	static const short offsets_from_0x4E00[] =
	{
		0,89,140,1345,1,1410,3,1372,186,869,875,93,992,831,134,1179,67,1024,1480,9,142,4210,143,2795,4069,2335,5197,3576,11,2855,8,14,7,10,
		3599,1507,4094,3697,1347,94,4061,191,42,1274,69,993,1482,2837,56,72,4223,161,18408,73,75,3640,4371,4081,4082,3408,1387,95,2931,19166,
		1027,96,1481,19564,97,48,9611,149,4352,2859,2857,6112,835,19,145,5198,13434,6440,148,5935,1413,13,2858,9388,1338,1414,3620,1483,1337,
		16742,4072,9305,3695,7636,146,1031,9958,7522,3601,6117,45,904,16157,901,7732,15297,1352,9307,5195,7643,7700,1351,18303,193,192,9287,
		198,1302,199,4097,205,197,6052,9258,1485,203,9270,206,202,1014,1030,79,876,211,6408,7695,1279,7456,19150,57,1280,76,996,1278,6023,877,
		6073,8811,58,6039,4550,15778,15777,5175,15780,4547,3642,4373,17,4084,3412,18463,1182,229,833,136,1181,1484,102,4219,9609,1034,11066,
		6443,6442,1019,5203,4071,7523,5201,5202,1183,5204,1467,10008,22,1508,13442,6444,6447,7751,2868,7753,3457,11380,3459,15784,921,15785,
		11068,15789,4549,15790,15791,15792,7736,1528,3644,7697,1018,7745,25,4070,1417,1523,10739,4099,20889,4211,8813,16743,28,1377,1303,1376,
		26,6119,4101,4434,20,6118,10478,1526,10034,10035,1518,10037,1527,10032,10033,1522,1514,2862,836,1532,1515,1510,1512,1529,2267,10015,
		2865,11198,24,216,215,227,217,236,234,10365,212,214,6053,9948,78,27,228,10024,10025,1392,80,1509,1286,908,9391,2838,2820,940,20511,
		1185,1285,19301,59,4098,11467,18410,880,1354,16829,2934,2932,1184,1516,10414,16825,1489,3413,2339,3577,1520,10715,12448,7629,4220,29,
		1519,4367,1041,1192,5211,3578,1545,5219,12803,5208,12801,5223,4073,2366,5225,5227,2352,5228,2362,12851,881,13458,154,13469,6461,6452,
		6458,6467,16839,13283,909,1359,15231,1419,1420,2344,6409,10366,3416,12812,19061,1312,2872,2874,8816,16830,1047,7547,5136,2873,16744,
		17066,1042,16840,7637,13299,7524,16158,4104,3608,3606,1187,841,4435,6121,1552,1555,14443,6386,2274,1548,1546,1539,2272,1592,1559,3711,
		4102,3713,2270,3714,1050,1049,12937,12625,4212,6449,840,34,13324,11513,16833,84,287,288,82,83,273,269,271,280,272,4342,246,251,292,
		247,253,1358,240,255,281,298,13290,14912,1553,316,1550,14924,13343,872,282,6464,1544,838,257,279,9271,286,1051,12940,6453,6466,1393,
		6124,6120,16159,1540,1549,2842,137,13426,2798,946,944,4228,4230,166,1048,20816,164,7457,14947,167,947,837,2948,18413,18414,18415,
		12682,4214,883,11891,8815,4062,7767,7777,7775,7776,7780,4569,884,3463,3464,3461,3415,3465,15794,923,15800,15802,924,15805,15806,15807,
		3579,17059,16837,3645,3580,4354,3417,18485,18483,5942,18486,18484,18482,2936,2946,2951,2941,2937,2952,5135,12733,15298,7458,112,12450,
		12452,12455,12454,12458,19568,4065,3583,4356,20646,4450,16859,5138,1566,16860,16861,16848,5238,5274,2395,5248,2383,5232,5330,5246,5241,
		5231,2368,16240,5252,2397,16161,5947,16228,5272,5267,5230,5282,3421,2375,5275,5269,2399,5271,2385,2378,5270,5284,2803,4567,5229,2391,
		2800,5258,5285,1396,1195,13501,13489,13497,13484,13517,13491,37,13478,1203,843,13519,6470,6496,6492,6480,6481,6479,6529,6478,6504,7746,
		6388,6495,15942,36,61,1339,16816,1201,1574,16856,7548,6501,16862,7525,2394,6129,10479,1608,6134,1588,1193,1471,17356,1606,2285,6135,2292,
		1600,1576,16307,17070,10039,2288,1589,50,1624,1580,1577,1593,1628,1575,1596,1067,3735,4112,16162,18056,18057,1610,5137,113,1065,11203,
		11200,11201,7631,885,304,339,341,326,312,348,303,310,355,334,352,335,333,308,16555,10370,347,16849,4475,4473,16852,345,4108,2384,15927,
		2981,1579,17083,3732,12957,12954,12960,20895,845,9410,9401,15314,1056,6497,1397,3739,16846,19309,19310,12027,15360,955,9398,169,949,4234,
		4243,10135,4244,951,16857,4239,16795,4355,950,4568,18418,18420,18423,1060,8822,8831,4383,7786,7833,7805,7811,7899,7839,7841,7816,7817,4608,
		4583,4587,3468,3467,3471,9314,11382,11383,8830,13423,15809,1583,15812,14949,1053,11070,15814,15817,15818,15821,15825,1563,8821,1395,3650,
		3647,3646,16863,3648,5945,4384,4556,18501,18502,18495,18504,18491,18500,2969,2966,2984,1194,4557,1202,20513,19569,12463,12465,12467,12466,
		19571,12469,12471,12472,12473,12474,19572,12477,2889,9641,9647,7526,18770,16163,9648,14952,15300,5305,5346,5332,5347,5317,2406,5308,5309,
		5328,5334,5325,12805,19062,5318,5349,5301,5336,1215,5297,2435,5321,5350,5324,4216,5339,2401,5291,5352,5353,5292,886,1494,13542,13541,13570,
		13561,13527,13553,13571,10484,13572,13582,13573,6551,6557,6511,6620,6544,6527,6526,6570,6532,6512,16880,6549,39,5142,10043,1383,139,1082,
		6563,18664,1366,10751,10753,1429,2900,2887,2891,4609,7463,2436,3003,16752,19063,16748,6057,16750,16751,1072,18782,1492,12975,20863,155,
		14414,14415,12990,16164,3610,6138,887,6556,1651,6150,2301,6156,10053,6158,6163,6146,888,2298,4576,1680,1660,20515,1679,1634,3768,3753,4118,
		12631,4124,3757,1007,16165,16169,16173,2302,18067,1078,10725,2434,9319,9321,86,1070,11206,1676,3427,3028,371,397,411,383,395,9288,388,422,
		423,1005,424,361,16167,413,10372,16875,16168,7459,4481,4480,9260,4476,4484,5184,13325,17361,1661,6055,9272,17351,1495,115,16170,4597,16171,
		12964,12986,12962,12991,12992,6411,12961,12965,6413,12993,1640,6159,20092,852,9424,4605,9431,2823,19312,19313,19314,1496,172,171,4247,2844,
		4249,4252,4245,1090,17098,4255,960,10482,5950,1083,12978,18424,18425,17105,1080,1399,1365,8850,8842,8853,8846,8841,7851,8005,7893,7876,7859,
		7870,7914,7865,7882,7871,7905,7912,7931,7923,7909,7864,7906,7932,7933,7867,4630,4647,4629,4636,4650,3430,3485,3479,3482,3484,3489,3481,3480,
		11386,4120,3486,15829,17102,15831,12969,5183,15834,14956,14955,15302,15837,15838,15842,15845,15846,4346,12931,4437,18614,3653,3658,1079,3656,
		4390,5247,3423,3428,18517,18509,18512,3001,3025,3024,3027,3019,19582,1474,13424,12479,12483,12484,12486,19574,12487,12488,19579,19580,12493,
		12495,16175,2895,6181,4142,9677,9659,7634,2443,5378,3585,5377,19065,2478,5390,2510,5408,5951,16244,16245,5409,5434,5356,5364,5374,5393,5383,
		2475,5411,5412,5372,5398,5385,5413,5418,6608,10010,18793,13648,4087,4134,13641,13607,13622,13650,13611,13665,13667,5957,13025,1367,13679,6663,
		6575,6596,6667,10488,6629,6607,6643,6641,6655,6671,6673,15233,1720,3073,7530,10772,10774,1432,1434,10764,10765,18786,12816,12813,9333,7563,
		7555,16763,20518,10374,13004,5144,8889,52,15304,11478,10497,1098,3613,6191,10492,10536,1732,6206,1745,914,6176,6175,6184,10063,16308,12995,
		16181,10060,14457,14462,14465,4637,14466,14461,1729,1725,19586,1751,1713,1741,1736,1708,1715,1770,8877,3809,12634,16177,16180,19624,18078,
		18079,18082,18085,18089,1400,12600,5340,10507,10729,4622,9330,16905,16898,11218,19353,11213,11211,11217,17357,2829,11519,7605,447,489,16183,
		19066,494,477,451,430,493,471,472,481,10375,7881,19772,437,16893,458,10494,4485,4491,4488,19067,1497,1105,16899,19167,10438,12998,13020,13022,
		13014,13065,1225,9453,9454,9452,9441,9457,9440,16184,4648,4645,19318,14464,19322,19324,4399,3590,2902,1728,173,174,4262,16889,4269,10158,10159,
		10155,10148,3071,178,18931,4125,6077,18427,18432,18433,4078,891,12686,3036,1499,16897,11899,16887,1101,19350,16902,4667,8892,8888,8878,8898,1091,
		7937,7978,7954,8007,8010,7966,8011,7959,7995,7998,7997,6611,8014,7947,7986,8017,8019,7973,4690,4706,4720,4732,4712,62,15305,3491,3492,3499,3498,
		11393,11391,11395,3490,928,15853,5185,14980,11094,11102,11101,15855,15857,15860,15861,2470,16896,6114,3659,6204,16185,18529,10505,3433,18532,
		18537,18530,3075,3045,3048,3067,3079,4626,6582,16186,10440,1223,4640,6612,2450,12497,12498,12499,12501,19588,12504,12505,12508,19590,12509,
		12510,12511,12821,12823,13427,7920,9696,9709,12064,14485,19069,10447,1322,5470,6717,5461,5423,16765,16246,16247,10448,5454,5455,2507,5449,5446,
		5456,5471,17149,1778,16925,5473,5474,5437,8941,4688,2806,5416,12859,12861,4717,13746,13739,13687,13751,6219,4726,10527,6726,6722,6755,6736,6698,
		6757,6723,6716,6689,6712,6711,6713,12066,1765,16927,16919,6679,17229,12741,16817,1799,2831,10816,10804,1439,2903,16912,8904,7562,19070,16767,
		16771,19071,7641,13300,6644,6732,14417,10449,12071,858,6226,10528,6227,20525,6211,6220,6229,14474,1768,1773,4713,1828,1866,1801,12642,3824,2310,
		16188,16191,18097,18099,18107,18113,18115,18117,12602,7719,9337,9338,16928,88,5964,11236,11231,11247,11239,11241,11248,11224,16911,11540,11537,
		11531,570,543,572,538,574,530,536,497,545,537,495,525,550,613,13293,3588,16556,4719,4498,4496,13360,13361,13356,13354,16916,5375,9273,9265,19074,
		12737,13062,13058,13048,13043,13071,13046,13073,9464,9468,16930,10073,10417,19327,4683,6760,8006,14960,19672,4141,966,4263,13066,10183,10181,
		10174,10172,10162,5960,11195,1808,16196,969,11481,1110,11486,17128,6081,6085,10076,18437,12702,9974,5363,11913,6041,10442,892,8932,8920,8934,
		8935,8923,8927,16914,8091,8025,8093,17234,8073,8072,8041,8055,8066,8052,8046,8001,8102,8042,8056,8104,8939,8076,4767,4740,4756,4774,3507,3517,
		3510,3509,3508,3518,11396,3513,3504,6728,15863,6423,15864,15867,5191,15004,14998,14989,15019,11109,15870,15873,15875,932,15877,15880,15882,1125,
		4723,3669,1127,3665,4401,18549,18550,18551,18538,3121,3096,16922,13053,18622,19076,6737,12514,12515,19596,12519,9731,9734,5479,2613,5583,2527,5545,
		5495,5522,5513,2566,5544,5504,5512,5977,5519,5536,2553,5541,5543,5538,5486,5528,12876,2554,13911,1234,20676,13836,13853,13772,13788,13828,13770,
		13837,13792,13861,6832,6822,6818,6789,6848,6835,6831,6774,5969,1135,11112,5146,9277,12875,15021,10459,18666,16773,16774,14426,18624,2562,4152,1305,
		6248,10561,10543,10556,4780,17358,1894,6234,1860,16349,16323,10085,14535,12079,1841,4771,1839,3862,3885,3847,2312,18140,18162,18166,10012,6824,9345,
		11259,11560,11580,11547,11558,11564,5967,602,14987,4768,639,630,631,4776,1838,604,591,583,4503,14932,10456,13369,6044,10450,20541,4745,7474,4457,
		19078,13082,13078,13112,13105,15969,2847,9500,9514,9486,9515,9499,19333,19334,977,975,7659,20667,10194,10197,4298,4279,4280,20607,10455,11488,11487,
		1862,6095,6091,6427,9607,10560,10454,11928,11927,11922,6061,1130,893,8197,8187,8139,8185,8224,8208,8183,8212,8152,8114,8161,8177,3142,6785,8215,
		4805,4828,4845,4796,4839,4821,4810,4840,4847,3527,3524,3519,11409,3526,15883,15886,11128,15900,16942,5986,3680,4409,18575,14539,18566,18576,3162,
		3190,19080,12521,12522,12525,19601,12531,12532,12533,12536,12543,9780,6033,6399,7486,2602,5677,2644,16266,16257,16267,16261,5584,2596,1370,5613,1948,
		5586,5610,5660,9070,5620,15041,5633,5634,5637,5601,5577,6063,6431,7482,12884,5987,4857,13932,13915,13923,13921,5996,13937,13885,6429,16796,13941,
		6866,6859,6925,6894,6917,6930,6901,6861,6857,6874,6869,4832,4817,16956,1448,1446,10860,10862,18625,7574,15042,18628,6274,18629,16776,4786,12075,16777,
		5982,16207,5516,6260,6289,6400,17359,1975,6262,1927,16967,1930,6255,16373,16332,16337,16983,14553,14555,14611,1949,1922,1944,1929,4165,4157,16204,
		16212,20689,18168,18170,18174,18176,18177,18180,18181,18184,18187,18192,10733,6266,7663,20549,1129,11277,11275,11264,11278,11600,11593,11601,11606,
		11611,11602,11604,11595,11613,690,645,9292,2593,18630,8998,653,680,2917,14935,4841,4513,4522,13383,13330,10090,17354,11197,13130,13118,13140,20097,
		9534,9524,9014,19339,15045,14574,3633,10203,11493,18452,1924,12705,6254,11946,3594,16979,6398,9008,8239,8278,8227,8319,8233,8244,8401,8318,8225,8248,
		8395,8329,4900,4940,4848,4903,4873,4968,1138,3538,3532,11420,11421,11415,16973,15061,15076,15065,15906,15907,15910,3678,3681,4410,11941,10127,18580,
		18585,12142,3266,10363,12558,12563,12566,19607,12568,9822,19778,12934,5700,5752,2667,5647,2636,20755,5702,5690,5676,5703,5662,2648,5706,13980,1252,
		20554,14045,2707,4181,14060,14020,13977,14008,9518,11137,7002,4851,7120,7046,7036,7042,16214,17260,4895,10893,10897,10894,10928,10903,10892,18679,18678,
		18686,18681,16787,10595,20868,17524,10587,10593,10604,17177,4890,6294,10463,7495,6295,9063,16360,16371,16362,16367,16351,16995,14590,14594,2003,12654,
		12650,12649,18201,18209,18211,18212,18214,18222,18223,10734,16798,11296,4865,11641,11646,11648,7617,13317,20768,684,699,719,16562,4526,4872,16997,13168,
		13157,13177,13182,13183,15334,15331,17265,10224,4297,6064,18933,4879,11950,6000,9038,2641,4936,9060,9036,8417,8480,8336,8420,8421,8404,8362,8348,8410,
		8424,11953,8425,4942,15497,2654,15912,11151,12708,7615,16799,18588,3276,3265,1504,12573,12576,18777,10919,9859,2713,5767,2057,5735,5162,15507,2691,5720,
		5716,12890,14141,4949,6318,14097,7201,7092,7068,7080,7500,17005,17271,17279,17272,10945,4927,18688,4362,15091,19095,2045,14651,14625,14663,14616,16218,
		18233,18235,13342,11315,11671,11689,11681,730,20795,19780,16012,13212,13194,13184,20124,10129,19346,15097,6002,15978,13199,17006,13136,10278,16803,11501,
		11503,6103,11966,7497,9092,9108,8454,8450,8491,8436,8468,8463,4962,3560,16219,3551,14620,15921,3305,12768,9098,1011,19617,12585,4967,5781,5778,16291,
		16287,5777,5805,5790,5796,2718,12906,18827,14153,14124,7210,7165,7217,7265,19160,17291,17289,18695,18697,10642,19096,6324,10638,4465,16418,16399,16425,
		16426,14710,14708,2097,2728,18247,18784,11323,20686,11327,11324,11697,11693,11719,757,16570,763,4535,13400,13213,13211,9119,5737,19100,7621,11978,17013,
		8540,8558,5058,19101,4976,1160,5837,9173,14255,14250,14212,19104,7256,6004,14733,17298,19216,2100,16452,2152,16224,20696,18268,16222,11758,17024,14945,
		13224,18645,10984,989,16808,16809,11990,11989,9155,8609,8640,5074,2753,17023,12596,5172,5862,18848,14287,18716,18718,10663,16456,14778,11351,12353,16811,
		16226,11999,12000,9189,13250,12796,19620,18861,15238,16486,18288,12795,20592,15718,5888,16498,19108,9955,9222,10118,2788,12800,16513,2236,2231,11853,19796,
		8780,14882,18744,18738,2250,12624, // End of 2,500 characters, append your own characters below.
	};
	static ImWchar base_ranges[] =
	{
		0x0020, 0x00FF, // Basic Latin + Latin Supplement
		0x3000, 0x30FF, // Punctuations, Hiragana, Katakana
		0x31F0, 0x31FF, // Katakana Phonetic Extensions
		0xFF00, 0xFFEF, // Half-width characters
	};
	static bool full_ranges_unpacked = false;
	static ImWchar full_ranges[IM_ARRAYSIZE(base_ranges) + IM_ARRAYSIZE(offsets_from_0x4E00) * 2 + 1];
	if (!full_ranges_unpacked)
	{
		// Unpack
		int codepoint = 0x4e00;
		memcpy(full_ranges, base_ranges, sizeof(base_ranges));
		ImWchar* dst = full_ranges + IM_ARRAYSIZE(base_ranges); // offset to skip base ranges.
		for (int n = 0; n < IM_ARRAYSIZE(offsets_from_0x4E00); n++, dst += 2)
			dst[0] = dst[1] = (ImWchar)(codepoint + offsets_from_0x4E00[n]); // code single character.
		dst[0] = 0;
		full_ranges_unpacked = true;
	}
	return &full_ranges[0];
}
```
**This function only load a small set of Chinese characters(about 7.14% of total 35,000 characters), but covers 97.97% most frequently used Chinese characters. If anybody find some character missing, just look up their Unicode and append to the list, and everything shall be fine. ;)**

**Also, this function reduces the size of font texture from 4kX8k to 2kX2k, about 40-50MB allocated memory, pretty affordable in most cases.**

Texture Preview:
![2500](https://user-images.githubusercontent.com/25875303/40967997-179832ac-68e7-11e8-80d0-d62d89c9b461.png)

**Comment by ocornut** [comment]
FYI, work on `dynamic_fonts` branch aimed to be merged in 1.92 (#8465) solves this, as glyphs are rasterized and packed on demand when they are used.
